masonry_winit/src/app_driver.rs

```diff
@@ -1,59 +1,161 @@
 // Copyright 2024 the Xilem Authors
 // SPDX-License-Identifier: Apache-2.0
 
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::num::NonZeroU64;
+use std::sync::atomic::{AtomicU64, Ordering};
+
+use masonry::core::Widget;
+use tracing::field::DisplayValue;
+use winit::event_loop::ActiveEventLoop;
+use winit::window::{Window as WindowHandle, WindowAttributes};
+
 use crate::app::{MasonryState, RenderRoot};
 use crate::core::{Action, WidgetId};
+use crate::event_loop_runner::WindowState;
 
-/// Context for the [`AppDriver`] trait.
+/// A unique and persistent identifier for a window.
 ///
-/// Currently holds a reference to the [`RenderRoot`].
-pub struct DriverCtx<'a> {
-    // We make no guarantees about the fields of this struct, but
-    // they must all be public so that the type can be constructed
-    // externally.
-    // This is needed for external users, whilst our external API
-    // is not yet designed.
-    #[doc(hidden)]
-    pub render_root: &'a mut RenderRoot,
+/// [`MasonryState`] internally maps these to winit window ids ([`winit::window::WindowId`]).
+/// Applications should only use this struct and not be concerned with the winit window ids.
+/// When the application is suspended and resumed this id will stay the same, while the
+/// winit window id will change.
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
+pub struct WindowId(pub(crate) NonZeroU64);
+
+impl WindowId {
+    /// Allocate a new, unique `WindowId`.
+    ///
+    /// You must ensure that a given `WindowId` is only ever used for one
+    /// window at a time.
+    pub fn next() -> Self {
+        static WINDOW_ID_COUNTER: AtomicU64 = AtomicU64::new(1);
+        let id = WINDOW_ID_COUNTER.fetch_add(1, Ordering::Relaxed);
+        Self(id.try_into().unwrap())
+    }
+
+    /// A serialized representation of the `WindowId` for debugging purposes.
+    pub fn trace(self) -> DisplayValue<NonZeroU64> {
+        tracing::field::display(self.0)
+    }
+}
+
+/// Context for the [`AppDriver`] trait.
+pub struct DriverCtx<'a, 's> {
+    state: &'a mut MasonryState<'s>,
+    event_loop: &'a ActiveEventLoop,
+}
```

Comment:
Types with two lifetime parameters are kind of a pain to use.

If there's any possibility to make this a one-lifetime type, I'd like that instead. If not, I'd want to at least document the reasons, maybe add a TODO comment.

Reply:
`&'a mut Something<'a>` is never what you want. Because of invariance the receiver is borrowed for the rest of its lifetime.

Reply:
Right, but usually the way we avoiding it is by avoiding referencing something that's a reference type. In our case, that could mean splitting away `MasonryState::surfaces`, for example.

Regardless, I'm not going to hold up the PR on it.

Reply:
The context can be used to create a new window including it's surface so splitting this up wouldn't be completely trivial. But yeah I agree we can do this later.